### ISSUE TITLE
noise accumulation by layer size 

### ISSUE THREAD

**Issue by ElinorG11** [IBM/aihwkit#344]
Hello,

I'm trying to check how the layer size of the network effects the noisiness of the output.

Technical details of the test:

- Architecture: ResNet34() structure for training on CIFAR from [here](https://github.com/kuangliu/pytorch-cifar/blob/master/models/resnet.py)
- Dataset: randomly generated values from gaussian distribution
- OS: Linux, Ubuntu 20.04
- GPU: Nvidia GeForce RTX 3070
- kit version: 0.4.0

in the test, which can be viewed [here](https://github.com/ElinorG11/Project-A/tree/main/test1_layer_size_location_noise_accumulation), I've used the following steps:

1. separated the resnet34 model into the most basic layers (conv2d, relu, batchnorm2d, etc...) 
2. converted each layer to analog and activated the noise as described [here](https://aihwkit.readthedocs.io/en/latest/pcm_inference.html#example-of-how-to-use-the-pcm-noise-model-for-inference)
3. fed the data into each conv2d/linear layer - to the ideal one and to the analog one and stored the outputs. I'll refer to the ideal output as the x vector and the analog output as the y vector
4. plotted the mean & std of the y-x vector 

the results are confusing, by looking at the graph attached below, it looks like there's no relation between the layer size and the accumulated noise.

![statistical_plot_by_layer_size_RESNET34](https://user-images.githubusercontent.com/48283282/148688110-42e0bdd5-6b5a-40e7-b769-749203abffcb.png)

It doesn't make any sense because  as you previously mentioned in other issues, the size of the layer determines the number of functioning devices on the crossbar, and I expect that as more devices are activated, the more noise I will see.

*** Note: the layer with (512) x-label is a linear layer. all the rest are convolutional layers.

I would appreciate it if you could confirm that those are indeed strange results, and perhaps explain in some more details where exactly in the code do you add the noise at each Crosspoint? perhaps you have any idea what could cause the strange pattern in the difference vector r=y-x?

Thank you very much for you time and attention.

**Comment by maljoras** [comment]
Hi @ElinorG11:

The noise induced deviation should scale proportional to `sqrt(n_x)`  where `n_x` is the number of input dimensions of a layer. The output dimension should not affect it.  What is computed for inference weight noise is basically `y_i = sum_j (w_ij+w_noise) x_j` thus `delta y_i = sum_j w_noise` (the deviation between ideal) which means that std of `y_delta` is proportional to `sqrt(n_x) std_wnoise`. 

However, there are also other noises and drift and the pcm noise is weight dependent, so the above is only approximate. For instance, there is also the `out_noise` turned on by default which is added to the output independently of the layer size. So you might need to turn it off to see the weight noise size dependence. Also noise and bound management might compromise the expected results.  

Besides, the real effect of the noise of the different layers is on the DNN output: you will notice that in this case of noisy (and small) first layers, the accuracy typically is most impacted.
